### ISSUE TITLE
[Bug] Rust rest client return parsing error for some transactions when calling `get_transaction_by_version`. ie. version: 523388616, 523363977

### ISSUE THREAD

**Issue by Alivers** [aptos-labs/aptos-core#12702]
# 🐛 Bug

The client entry returns error without a valid reason. There are many transactions with this error. 

## To reproduce

rust sdk rev is `branch = devnet`.
**Code snippet to reproduce**
```rust
 let client = RestClient::new(Url::parse("https://api.mainnet.aptoslabs.com/v1").unwrap());
print!("{:?}", client.get_transaction_by_version(523363977).await);
return Ok(());
```

**Stack trace/error message**
```
Err(Unknown(error decoding response body: invalid value: map, expected map with a single key

Caused by:
    invalid value: map, expected map with a single key))  
```

**Comment by Alivers** [comment]
Seem these transaction are all `MultiSig` transactions.  #12445 
After changing the sdk rev to `branch = main`. The error also changed to the below: 
```
Err(Unknown(error decoding response body: missing field `type`

Caused by:
    missing field `type`))      
```

**Comment by junkil-park** [comment]
@Alivers , could you confirm if this functionality in question was operational fine prior my PR (https://github.com/aptos-labs/aptos-core/pull/12445)?

I am asking because the multisig payload (de)serialization in JSON has been already broken, so my PR fixes it. The fix isn't rolled out to the devnet and the mainnet yet.

**Comment by junkil-park** [comment]
@Alivers , can you use the BCS version: `get_transaction_by_version_bcs` instead of `get_transaction_by_version`? BCS is more robust to this kind of change in general.

**Comment by Alivers** [comment]
> @Alivers , could you confirm if this functionality in question was operational fine prior my PR (https://github.com/aptos-labs/aptos-core/pull/12445)?
> 
> 
> 
> I am asking this because the multisig payload (de)serialization in JSON has been already broken, so my PR fixes it. The fix isn't rolled out to the devnet and the mainnet yet.

I’ve tried the fix in PR #12445, using the main branch, but I’m still encountering issues. The error description is mentioned in my first comment. Could you please take a look and provide some guidance or assistance?

**Comment by Alivers** [comment]
> @Alivers , can you use the BCS version: `get_transaction_by_version_bcs` instead of `get_transaction_by_version`? BCS is more robust to this kind of change in general.

This function works. But I need to change the Transaction Type Reference all my codebase, or is there any convenient way to  convert `TransactionData` to `Transaction`?

**Comment by junkil-park** [comment]
> I’ve tried the fix in PR https://github.com/aptos-labs/aptos-core/pull/12445, using the main branch, but I’m still encountering issues. The error description is mentioned in my first comment. Could you please take a look and provide some guidance or assistance?

My question is: Did get_transaction_by_version for a multisig transaction work previously? Has it only recently stopped functioning?

**Comment by junkil-park** [comment]
> This function works. But I need to change the Transaction Type Reference all my codebase, or is there any convenient way to convert TransactionData to Transaction?

Unfortunately, the conversion function doesn't seem to exist. However, the information contained in the two struct should largely overlap.

**Comment by Alivers** [comment]
> > I’ve tried the fix in PR #12445, using the main branch, but I’m still encountering issues. The error description is mentioned in my first comment. Could you please take a look and provide some guidance or assistance?
> 
> My question is: Did `get_transaction_by_version` for multisig transactions work previously? Has it only recently stopped functioning?

Did work for a long time. I just recently noticed that.

**Comment by Alivers** [comment]
> > This function works. But I need to change the Transaction Type Reference all my codebase, or is there any convenient way to convert TransactionData to Transaction?
> 
> Unfortunately, the conversion function doesn't seem to exist. However, the information contained in the two struct should largely overlap.
Ok, but it is too difficult for me to change the type and match multi-layer invariants(Transaction, Event, WriteSet).  Could you please fix this issue? 🙏🏻

**Comment by junkil-park** [comment]
> > > I’ve tried the fix in PR #12445, using the main branch, but I’m still encountering issues. The error description is mentioned in my first comment. Could you please take a look and provide some guidance or assistance?
> > 
> > 
> > My question is: Did `get_transaction_by_version` for multisig transactions work previously? Has it only recently stopped functioning?
> 
> Did work for a long time. I just recently noticed that.

I think that this has not been functioning on the mainnet. I tested it with `branch = mainnet`, and found that it also produced an error:
```
Err(Unknown(error decoding response body: invalid value: map, expected map with a single key

Caused by:
    invalid value: map, expected map with a single key))
```
Probably, you haven't encountered this Multisig payload cases. The transactions 523388616 and 523363977 are from very recent days, only a few days ago.

**Comment by junkil-park** [comment]
> > > This function works. But I need to change the Transaction Type Reference all my codebase, or is there any convenient way to convert TransactionData to Transaction?
> > 
> > 
> > Unfortunately, the conversion function doesn't seem to exist. However, the information contained in the two struct should largely overlap.
> 
> Ok, but it is too difficult for me to change the type and match multi-layer invariants(Transaction, Event, WriteSet). Could you please fix this issue? 🙏🏻

I believe that this PR (https://github.com/aptos-labs/aptos-core/pull/12445) is the fix to the issue you describe. The fix has already landed on the `main` branch, which needs to roll out to devnet -> testnet -> mainnet. This process will probably take 1.5 months from now.

I would highly recommend you to try the BCS version, and let me know if you have any question on that.

**Comment by Alivers** [comment]
> > > > I’ve tried the fix in PR #12445, using the main branch, but I’m still encountering issues. The error description is mentioned in my first comment. Could you please take a look and provide some guidance or assistance?
> 
> > > 
> 
> > > 
> 
> > > My question is: Did `get_transaction_by_version` for multisig transactions work previously? Has it only recently stopped functioning?
> 
> > 
> 
> > Did work for a long time. I just recently noticed that.
> 
> 
> 
> I think that this has not been functioning on the mainnet. I tested it with `branch = mainnet`, and found that it also produced an error:
> 
> ```
> 
> Err(Unknown(error decoding response body: invalid value: map, expected map with a single key
> 
> 
> 
> Caused by:
> 
>     invalid value: map, expected map with a single key))
> 
> ```
> 
> Probably, you haven't encountered this Multisig payload cases. The transactions 523388616 and 523363977 you mention are from very recent days, only a few days ago.

The mainnet branch produced the invalid value map error. But the main branch(which contains the changes of #12445) prodeces the type field missing error. Two different branches.

**Comment by junkil-park** [comment]
> > > > > I’ve tried the fix in PR #12445, using the main branch, but I’m still encountering issues. The error description is mentioned in my first comment. Could you please take a look and provide some guidance or assistance?
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > My question is: Did `get_transaction_by_version` for multisig transactions work previously? Has it only recently stopped functioning?
> > 
> > 
> > > 
> > 
> > 
> > > Did work for a long time. I just recently noticed that.
> > 
> > 
> > I think that this has not been functioning on the mainnet. I tested it with `branch = mainnet`, and found that it also produced an error:
> > ```
> > 
> > Err(Unknown(error decoding response body: invalid value: map, expected map with a single key
> > 
> > 
> > 
> > Caused by:
> > 
> >     invalid value: map, expected map with a single key))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Probably, you haven't encountered this Multisig payload cases. The transactions 523388616 and 523363977 you mention are from very recent days, only a few days ago.
> 
> The mainnet branch produced the invalid value map error. But the main branch(which contains the changes of #12445) prodeces the type field missing error. Two different branches. Not sure if it was caused by the multisig tx parsing.

It's not just a parsing (deserialization) issue but also a serialization issue. More specifically, the error is caused by the misalignment of serialization and deserialization of MultisigTransactionPayload in JSON, which has exists for a while (https://github.com/aptos-labs/aptos-core/issues/8304).

My fix #12445 makes them aligned. It's expected that even though you use `aptos-sdk` from `branch = main`, you still encounter the error, because the mainnet fullnode currently serializes those payloads incorrectly (or in a misaligned way). So, the issue will be resolved once the fix rolls out to the mainnet fullnode.

**Comment by Alivers** [comment]
> > > > > > I’ve tried the fix in PR #12445, using the main branch, but I’m still encountering issues. The error description is mentioned in my first comment. Could you please take a look and provide some guidance or assistance?
> 
> > > 
> 
> > > 
> 
> > > > >
> 
> > > 
> 
> > > 
> 
> > > > >
> 
> > > 
> 
> > > 
> 
> > > > > My question is: Did `get_transaction_by_version` for multisig transactions work previously? Has it only recently stopped functioning?
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > Did work for a long time. I just recently noticed that.
> 
> > > 
> 
> > > 
> 
> > > I think that this has not been functioning on the mainnet. I tested it with `branch = mainnet`, and found that it also produced an error:
> 
> > > ```
> 
> > > 
> 
> > > Err(Unknown(error decoding response body: invalid value: map, expected map with a single key
> 
> > > 
> 
> > > 
> 
> > > 
> 
> > > Caused by:
> 
> > > 
> 
> > >     invalid value: map, expected map with a single key))
> 
> > > ```
> 
> > > 
> 
> > > 
> 
> > >     
> 
> > >       
> 
> > >     
> 
> > > 
> 
> > >       
> 
> > >     
> 
> > > 
> 
> > >     
> 
> > >   
> 
> > > Probably, you haven't encountered this Multisig payload cases. The transactions 523388616 and 523363977 you mention are from very recent days, only a few days ago.
> 
> > 
> 
> > The mainnet branch produced the invalid value map error. But the main branch(which contains the changes of #12445) prodeces the type field missing error. Two different branches. Not sure if it was caused by the multisig tx parsing.
> 
> 
> 
> It's not just a parsing (deserialization) issue but also a serialization issue. More specifically, the error is caused by the misalignment of serialization and deserialization of MultisigTransactionPayload in JSON, which has exists for a while (https://github.com/aptos-labs/aptos-core/issues/8304).
> 
> 
> 
> My fix #12445 makes them aligned. It's expected that even though you use `aptos-sdk` from `branch = main` now, you still encounter the error, because currently, the mainnet fullnode incorrectly serializes those payloads (or in a misaligned way). So, the issue will be resolved once the fix rolls out to the mainnet fullnode.

Got it! Thank you.
